File: plugins/project-intelligence/scripts/project_intel_lib/quality.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


EXCLUDED_PACKAGE_PARTS = {"node_modules", "dist", "build", ".git", ".project-intel", ".next", ".nuxt"}


def load_json(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}
    return payload if isinstance(payload, dict) else {}


def workspace_package_files(root: Path, root_package: dict[str, Any]) -> list[Path]:
    workspaces = root_package.get("workspaces", [])
    if isinstance(workspaces, dict):
        workspaces = workspaces.get("packages", [])
    patterns = [item for item in workspaces if isinstance(item, str)] if isinstance(workspaces, list) else []
    candidates: set[Path] = set()
    for pattern in patterns:
        normalized = pattern.rstrip("/")
        for directory in root.glob(normalized):
            path = directory / "package.json" if directory.is_dir() else directory
            if path.name == "package.json" and path.is_file():
                candidates.add(path)
    if not patterns:
        for parent in ("apps", "packages", "services", "frontend", "backend"):
            base = root / parent
            if not base.is_dir():
                continue
            for path in base.glob("*/package.json"):
                candidates.add(path)
    return sorted(
        (path for path in candidates if not any(part in EXCLUDED_PACKAGE_PARTS for part in path.relative_to(root).parts)),
        key=lambda path: path.as_posix(),
    )


def package_frameworks(deps: dict[str, Any]) -> list[str]:
    frameworks: list[str] = []
    for name, markers in {
        "Vue": ["vue", "@vitejs/plugin-vue", "nuxt"],
        "React": ["react", "next", "@vitejs/plugin-react"],
        "Vite": ["vite"],
        "NestJS": ["@nestjs/core", "@nestjs/common"],
        "Express": ["express"],
        "Koa": ["koa"],
        "Fastify": ["fastify"],
        "TypeScript": ["typescript"],
        "Element Plus": ["element-plus"],
        "Pinia": ["pinia"],
        "Vue Router": ["vue-router"],
        "Redux": ["redux", "@reduxjs/toolkit"],
    }.items():
        if any(marker in deps for marker in markers):
            frameworks.append(name)
    return frameworks


def detect_package(root: Path) -> dict[str, Any]:
    root_payload = load_json(root / "package.json")
    package_files = ([root / "package.json"] if root_payload else []) + workspace_package_files(root, root_payload)
    packages: list[dict[str, Any]] = []
    frameworks: set[str] = set()
    for path in package_files:
        payload = root_payload if path == root / "package.json" else load_json(path)
        deps: dict[str, Any] = {}
        for key in ("dependencies", "devDependencies", "peerDependencies"):
            deps.update(payload.get(key, {}) or {})
        detected = package_frameworks(deps)
        frameworks.update(detected)
        packages.append(
            {
                "path": path.parent.relative_to(root).as_posix() or ".",
                "name": payload.get("name"),
                "scripts": payload.get("scripts", {}) or {},
                "frameworks": detected,
            }
        )
    for name, markers in {
        "Java/Spring": ["pom.xml", "build.gradle", "build.gradle.kts"],
        "Python": ["pyproject.toml", "requirements.txt", "manage.py"],
        "Go": ["go.mod"],
        "Rust": ["Cargo.toml"],
    }.items():
        if any((root / marker).exists() for marker in markers):
            frameworks.add(name)
    return {
        "packageName": root_payload.get("name"),
        "scripts": root_payload.get("scripts", {}) or {},
        "frameworks": sorted(frameworks),
        "hasPackageJson": bool(root_payload),
        "packages": packages,
        "workspace": len(packages) > 1,
    }


def package_manager(root: Path) -> str:
    if (root / "pnpm-lock.yaml").exists():
        return "pnpm"
    if (root / "yarn.lock").exists():
        return "yarn"
    return "npm"


def package_script_command(manager: str, directory: str, script: str) -> str:
    if directory in {"", "."}:
        return f"{manager} run {script}"
    if manager == "pnpm":
        return f"pnpm --dir {directory} run {script}"
    if manager == "yarn":
        return f"yarn --cwd {directory} run {script}"
    return f"npm --prefix {directory} run {script}"


def append_unique(commands: list[dict[str, Any]], item: dict[str, Any]) -> None:
    if item.get("command") and not any(existing.get("command") == item.get("command") for existing in commands):
        commands.append(item)


def detect_quality_commands(root: Path, package: dict[str, Any]) -> list[dict[str, Any]]:
    commands: list[dict[str, Any]] = []
    aliases = {
        "lint": ["lint", "lint:eslint", "eslint"],
        "type-check": ["type-check", "typecheck", "check-types", "vue-tsc", "tsc"],
        "format-check": ["format:check", "prettier:check", "check:format"],
        "style-check": ["stylelint", "lint:style", "style:check"],
        "test": ["test", "test:unit", "unit"],
    }
    manager = package_manager(root)
    packages = package.get("packages", []) or ([{"path": ".", "scripts": package.get("scripts", {})}] if package.get("hasPackageJson") else [])
    for workspace in packages:
        directory = workspace.get("path") or "."
        scripts = workspace.get("scripts", {}) or {}
        for kind, names in aliases.items():
            for name in names:
                if name in scripts:
                    append_unique(commands, {
                        "kind": kind,
                        "command": package_script_command(manager, directory, name),
                        "source": "package.json" if directory == "." else f"{directory}/package.json",
                    })
                    break

    root_kinds = {item.get("kind") for item in commands if item.get("source") == "package.json"}
    if "lint" not in root_kinds and any((root / name).exists() for name in ("eslint.config.js", "eslint.config.mjs", ".eslintrc.js", ".eslintrc.cjs", ".eslintrc.json")):
        append_unique(commands, {"kind": "lint", "command": "npx eslint .", "source": "inferred"})
    if "type-check" not in root_kinds and (root / "tsconfig.json").exists():
        command = "npx vue-tsc --noEmit" if "Vue" in package.get("frameworks", []) else "npx tsc --noEmit"
        append_unique(commands, {"kind": "type-check", "command": command, "source": "inferred"})
    if "style-check" not in root_kinds and any((root / name).exists() for name in ("stylelint.config.js", "stylelint.config.cjs", ".stylelintrc", ".stylelintrc.js")):
        append_unique(commands, {"kind": "style-check", "command": 'npx stylelint "**/*.{css,scss,vue}"', "source": "inferred"})
    if "format-check" not in root_kinds and any((root / name).exists() for name in ("prettier.config.js", "prettier.config.cjs", ".prettierrc", ".prettierrc.js", ".prettierrc.json")):
        append_unique(commands, {"kind": "format-check", "command": "npx prettier --check .", "source": "inferred"})

    if (root / "pom.xml").exists():
        mvn = "./mvnw" if (root / "mvnw").exists() else "mvn"
        append_unique(commands, {"kind": "test", "command": f"{mvn} test", "source": "pom.xml"})
        append_unique(commands, {"kind": "verify", "command": f"{mvn} verify", "source": "pom.xml"})
    if (root / "build.gradle").exists() or (root / "build.gradle.kts").exists():
        gradle = "./gradlew" if (root / "gradlew").exists() else "gradle"
        append_unique(commands, {"kind": "test", "command": f"{gradle} test", "source": "gradle"})
        append_unique(commands, {"kind": "verify", "command": f"{gradle} check", "source": "gradle"})
    if (root / "pyproject.toml").exists() or (root / "requirements.txt").exists():
        pyproject = (root / "pyproject.toml").read_text(encoding="utf-8", errors="ignore") if (root / "pyproject.toml").exists() else ""
        requirements = (root / "requirements.txt").read_text(encoding="utf-8", errors="ignore") if (root / "requirements.txt").exists() else ""
        combined = (pyproject + "\n" + requirements).lower()
        if "pytest" in combined or (root / "tests").is_dir():
            append_unique(commands, {"kind": "test", "command": "python3 -m pytest", "source": "python-project"})
        if "ruff" in combined:
            append_unique(commands, {"kind": "lint", "command": "python3 -m ruff check .", "source": "python-project"})
        if "mypy" in combined:
            append_unique(commands, {"kind": "type-check", "command": "python3 -m mypy .", "source": "python-project"})
    if (root / "go.mod").exists():
        append_unique(commands, {"kind": "test", "command": "go test ./...", "source": "go.mod"})
        append_unique(commands, {"kind": "verify", "command": "go vet ./...", "source": "go.mod"})
    if (root / "Cargo.toml").exists():
        append_unique(commands, {"kind": "type-check", "command": "cargo check", "source": "Cargo.toml"})
        append_unique(commands, {"kind": "test", "command": "cargo test", "source": "Cargo.toml"})
        append_unique(commands, {"kind": "format-check", "command": "cargo fmt -- --check", "source": "Cargo.toml"})
    return commands

File: src/scanner/quality.ts
// Package + quality-command detection (phase 3.A.4), ported from quality.py.
//
// Per the migration requirement (P1.3): the product itself must not depend on
// Python, but scanning a Python repository must STILL recognize pytest/ruff/mypy
// as the scanned project's quality commands. The detected command strings are
// the scanned project's commands (run by that project's toolchain), not the
// product's runtime — so detecting `pytest`/`ruff`/`mypy` is correct and
// preserved. The product's own runtime never invokes these.

import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { dirname, join, relative } from "node:path";

export const EXCLUDED_PACKAGE_PARTS = new Set([
  "node_modules",
  "dist",
  "build",
  ".git",
  ".project-intel",
  ".next",
  ".nuxt",
]);

function loadJson(path: string): Record<string, unknown> {
  try {
    const payload = JSON.parse(readFileSync(path, "utf8"));
    return payload && typeof payload === "object" && !Array.isArray(payload) ? payload : {};
  } catch {
    return {};
  }
}

function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function globPackageJson(dir: string): string[] {
  try {
    return readdirSync(dir)
      .map((name) => join(dir, name, "package.json"))
      .filter((p) => {
        try {
          return statSync(p).isFile();
        } catch {
          return false;
        }
      });
  } catch {
    return [];
  }
}

export function workspacePackageFiles(root: string, rootPackage: Record<string, unknown>): string[] {
  const workspacesRaw = rootPackage.workspaces;
  let patterns: string[] = [];
  if (Array.isArray(workspacesRaw)) {
    patterns = workspacesRaw.filter((s): s is string => typeof s === "string");
  } else if (workspacesRaw && typeof workspacesRaw === "object") {
    const pkg = (workspacesRaw as Record<string, unknown>).packages;
    if (Array.isArray(pkg)) {
      patterns = pkg.filter((s): s is string => typeof s === "string");
    }
  }
  const candidates = new Set<string>();
  for (const pattern of patterns) {
    const normalized = pattern.replace(/\/$/, "");
    // Support simple glob like "packages/*"
    if (normalized.includes("*")) {
      const base = normalized.split("*")[0]!.replace(/\/$/, "");
      const dir = join(root, base);
      if (isDir(dir)) {
        for (const pkg of globPackageJson(dir)) candidates.add(pkg);
      }
    } else {
      const dir = join(root, normalized);
      const pkg = join(dir, "package.json");
      if (existsSync(pkg)) candidates.add(pkg);
    }
  }
  if (!patterns.length) {
    for (const parent of ["apps", "packages", "services", "frontend", "backend"]) {
      const base = join(root, parent);
      if (isDir(base)) for (const pkg of globPackageJson(base)) candidates.add(pkg);
    }
  }
  return [...candidates]
    .filter((p) => {
      const rel = p.slice(root.length + 1).split("\\").join("/").split("/");
      return !rel.some((part) => EXCLUDED_PACKAGE_PARTS.has(part));
    })
    .sort();
}

const FRAMEWORK_MARKERS: Record<string, string[]> = {
  Vue: ["vue", "@vitejs/plugin-vue", "nuxt"],
  React: ["react", "next", "@vitejs/plugin-react"],
  Vite: ["vite"],
  NestJS: ["@nestjs/core", "@nestjs/common"],
  Express: ["express"],
  Koa: ["koa"],
  Fastify: ["fastify"],
  TypeScript: ["typescript"],
  "Element Plus": ["element-plus"],
  Pinia: ["pinia"],
  "Vue Router": ["vue-router"],
  Redux: ["redux", "@reduxjs/toolkit"],
};

export function packageFrameworks(deps: Record<string, unknown>): string[] {
  const frameworks: string[] = [];
  for (const [name, markers] of Object.entries(FRAMEWORK_MARKERS)) {
    if (markers.some((marker) => marker in deps)) frameworks.push(name);
  }
  return frameworks;
}

export function detectPackage(root: string): Record<string, unknown> {
  const rootPayload = loadJson(join(root, "package.json"));
  const rootHas = Object.keys(rootPayload).length > 0;
  const packageFiles = (rootHas ? [join(root, "package.json")] : []).concat(workspacePackageFiles(root, rootPayload));
  const packages: Record<string, unknown>[] = [];
  const frameworks = new Set<string>();
  for (const path of packageFiles) {
    const payload = path === join(root, "package.json") && rootHas ? rootPayload : loadJson(path);
    const deps: Record<string, unknown> = {};
    for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
      Object.assign(deps, (payload[key] as Record<string, unknown>) ?? {});
    }
    const detected = packageFrameworks(deps);
    detected.forEach((f) => frameworks.add(f));
    packages.push({
      path: relative(root, dirname(path)).replace(/\\/g, "/") || ".",
      name: payload.name,
      scripts: (payload.scripts as Record<string, unknown>) ?? {},
      frameworks: detected,
    });
  }
  const nonJsMarkers: Record<string, string[]> = {
    "Java/Spring": ["pom.xml", "build.gradle", "build.gradle.kts"],
    Python: ["pyproject.toml", "requirements.txt", "manage.py"],
    Go: ["go.mod"],
    Rust: ["Cargo.toml"],
  };
  for (const [name, markers] of Object.entries(nonJsMarkers)) {
    if (markers.some((m) => existsSync(join(root, m)))) frameworks.add(name);
  }
  return {
    packageName: rootPayload.name,
    scripts: (rootPayload.scripts as Record<string, unknown>) ?? {},
    frameworks: [...frameworks].sort(),
    hasPackageJson: rootHas,
    packages,
    workspace: packages.length > 1,
  };
}

export function packageManager(root: string): string {
  if (existsSync(join(root, "pnpm-lock.yaml"))) return "pnpm";
  if (existsSync(join(root, "yarn.lock"))) return "yarn";
  return "npm";
}

export function packageScriptCommand(manager: string, directory: string, script: string): string {
  if (directory === "" || directory === ".") return `${manager} run ${script}`;
  if (manager === "pnpm") return `pnpm --dir ${directory} run ${script}`;
  if (manager === "yarn") return `yarn --cwd ${directory} run ${script}`;
  return `npm --prefix ${directory} run ${script}`;
}

function appendUnique(commands: Record<string, unknown>[], item: Record<string, unknown>): void {
  if (item.command && !commands.some((c) => c.command === item.command)) commands.push(item);
}

export function detectQualityCommands(root: string, pkg: Record<string, unknown>): Record<string, unknown>[] {
  const commands: Record<string, unknown>[] = [];
  const aliases: Record<string, string[]> = {
    lint: ["lint", "lint:eslint", "eslint"],
    "type-check": ["type-check", "typecheck", "check-types", "vue-tsc", "tsc"],
    "format-check": ["format:check", "prettier:check", "check:format"],
    "style-check": ["stylelint", "lint:style", "style:check"],
    test: ["test", "test:unit", "unit"],
  };
  const manager = packageManager(root);
  const packages = (pkg.packages as Record<string, unknown>[]) ?? (pkg.hasPackageJson ? [{ path: ".", scripts: pkg.scripts }] : []);
  for (const workspace of packages) {
    const directory = (workspace.path as string) || ".";
    const scripts = (workspace.scripts as Record<string, unknown>) ?? {};
    for (const [kind, names] of Object.entries(aliases)) {
      for (const name of names) {
        if (name in scripts) {
          appendUnique(commands, {
            kind,
            command: packageScriptCommand(manager, directory, name),
            source: directory === "." ? "package.json" : `${directory}/package.json`,
          });
          break;
        }
      }
    }
  }
  const rootKinds = new Set(commands.filter((c) => c.source === "package.json").map((c) => c.kind));
  if (!rootKinds.has("lint") && ["eslint.config.js", "eslint.config.mjs", ".eslintrc.js", ".eslintrc.cjs", ".eslintrc.json"].some((n) => existsSync(join(root, n)))) {
    appendUnique(commands, { kind: "lint", command: "npx eslint .", source: "inferred" });
  }
  if (!rootKinds.has("type-check") && existsSync(join(root, "tsconfig.json"))) {
    const command = (pkg.frameworks as string[] | undefined)?.includes("Vue") ? "npx vue-tsc --noEmit" : "npx tsc --noEmit";
    appendUnique(commands, { kind: "type-check", command, source: "inferred" });
  }
  if (!rootKinds.has("style-check") && ["stylelint.config.js", "stylelint.config.cjs", ".stylelintrc", ".stylelintrc.js"].some((n) => existsSync(join(root, n)))) {
    appendUnique(commands, { kind: "style-check", command: 'npx stylelint "**/*.{css,scss,vue}"', source: "inferred" });
  }
  if (!rootKinds.has("format-check") && ["prettier.config.js", "prettier.config.cjs", ".prettierrc", ".prettierrc.js", ".prettierrc.json"].some((n) => existsSync(join(root, n)))) {
    appendUnique(commands, { kind: "format-check", command: "npx prettier --check .", source: "inferred" });
  }

  // Java
  if (existsSync(join(root, "pom.xml"))) {
    const mvn = existsSync(join(root, "mvnw")) ? "./mvnw" : "mvn";
    appendUnique(commands, { kind: "test", command: `${mvn} test`, source: "pom.xml" });
    appendUnique(commands, { kind: "verify", command: `${mvn} verify`, source: "pom.xml" });
  }
  if (existsSync(join(root, "build.gradle")) || existsSync(join(root, "build.gradle.kts"))) {
    const gradle = existsSync(join(root, "gradlew")) ? "./gradlew" : "gradle";
    appendUnique(commands, { kind: "test", command: `${gradle} test`, source: "gradle" });
    appendUnique(commands, { kind: "verify", command: `${gradle} check`, source: "gradle" });
  }
  // Python — STILL recognize pytest/ruff/mypy as the scanned project's commands (P1.3).
  if (existsSync(join(root, "pyproject.toml")) || existsSync(join(root, "requirements.txt"))) {
    let combined = "";
    try {
      if (existsSync(join(root, "pyproject.toml"))) combined += readFileSync(join(root, "pyproject.toml"), "utf8");
    } catch {
      /* ignore */
    }
    try {
      if (existsSync(join(root, "requirements.txt"))) combined += "\n" + readFileSync(join(root, "requirements.txt"), "utf8");
    } catch {
      /* ignore */
    }
    const lower = combined.toLowerCase();
    if (lower.includes("pytest") || isDir(join(root, "tests"))) {
      appendUnique(commands, { kind: "test", command: "pytest", source: "python-project" });
    }
    if (lower.includes("ruff")) {
      appendUnique(commands, { kind: "lint", command: "ruff check .", source: "python-project" });
    }
    if (lower.includes("mypy")) {
      appendUnique(commands, { kind: "type-check", command: "mypy .", source: "python-project" });
    }
  }
  if (existsSync(join(root, "go.mod"))) {
    appendUnique(commands, { kind: "test", command: "go test ./...", source: "go.mod" });
    appendUnique(commands, { kind: "verify", command: "go vet ./...", source: "go.mod" });
  }
  if (existsSync(join(root, "Cargo.toml"))) {
    appendUnique(commands, { kind: "type-check", command: "cargo check", source: "Cargo.toml" });
    appendUnique(commands, { kind: "test", command: "cargo test", source: "Cargo.toml" });
    appendUnique(commands, { kind: "format-check", command: "cargo fmt -- --check", source: "Cargo.toml" });
  }
  return commands;
}
